Create config single definition

alanoe opened the pull request:
Depends on builds/#200 and builds/#207

Commit:
Remove default values from host_os.py's command line arguments

Commit:
Improve merging of config file attributes and command line arguments in a dictionary

This also fixes unit tests which validate default values to not parse only
command line arguments, but config file too.

Commit:
Support YAML config file dict nodes overriding with command line arguments

Command line argument must be passed in JSON format.

Commit:
Remove mock_config attribute from config YAML

It can be inferred from Linux distro name, version, arch
and endianness.

Commit:
Make layout fixes in config.yaml

- Make indentation consistent
- Set indentation to 2 characters (next patch will create config.yaml from
  a YAML in RAM and PyYAML require a minimum indentation of 2 characters)
- Remove extra whitespaces

Commit:
Create config YAML file and command line parser from config metadata file

farosas commented on config_metadata.yaml:
```diff
@@ -0,0 +1,169 @@
+common:
```
Should this file be hidden to avoid inadvertent edits?

alanoe replied:
I don't know if it is necessary, given that the file content YAML structure is different from config.yaml. Alternatively, there could be a warning at the top of the file. Up to you guys.

farosas commented on lib/config.py:
```diff
@@ -155,47 +46,61 @@ def __init__(self):
-                for arg, options in arg_group.items():
-                    command_parser.add_argument(*arg, **options)
+
+        # setup parser from config metadata file
```
Why do we need to infer type, nargs, action? Wasn't that the supposition when we *didn't* have a config_metadata file? Why not simply keep these explicitly as well as any other `add_argument` argument? I think it would simplify the code below, avoid any bugs in the inferring logic and also allow us to use other arguments.

Or am I forgetting something here?

alanoe replied:
I am not sure I understood you. Default action is 'store', somewhere we must explicitly define the action if it is a different value (eg store_true or store_false). Same for nargs. The difference is that before we had those things defined for *each argument* in the SUBCOMMANDS constants, now they are inferred for *all arguments* from the config metadata file. The idea here is to automatically create the command line parser from the config metadata file, there should be no code specific to one argument.

farosas replied:
My suggestion is to explicitly set these values in config_metadata now that this won't be edited by the user:
```
verbose: 
  argparse_args:
    help:
    default:
    action:
    const:
    choices:
    required:
    metavar: 
    dest:
```

Just bringing to your attention that the current design was based on premises that are not true anymore, that's all.

alanoe replied:
I vote for the design proposed in this patch, which infers most of these argparse fields and requires us to keep a simpler and smaller config metadata file. I still do not understand the benefit of mentioning those explicitly.

olavphibm replied:
I agree with @fabianorosas that setting all values in the config_metadata file could make the code less error prone and the configuration more flexible. This flexibility could be useful in the future. I do think that defaults could be left out to avoid making the file too big, however. The argparse defaults could be set like this, for example: http://stackoverflow.com/a/12627202

lisibann reviewed (commented):
We need some documentation related to the config metadata file and the execution of sync_config_yaml.py.

farosas reviewed (approved):
Reviewed only b5bf78a28f7999e044deede23d247f52108a7e1a as requested

olavphibm commented on config.yaml:
```diff
@@ -1,88 +1,85 @@
+build_iso:
+  automated_install_file: host-os.ks
+  automated_install_packages_groups:
+  - '@base'
```
Shouldn't this be at a higher indentation level?

alanoe replied:
This is the default serialization behavior in PyYAML and yamllint does not complain.

olavphibm commented on lib/config.py:
```diff
@@ -149,46 +49,61 @@ def __init__(self):
-                                 help='Directory used to store all temporary '
-                                 'files created during the process.')
+                                      'scripts',
+                                 # NOTE(maurosr): move this to /etc in the future
```
I believe this note could be removed.

alanoe replied:
I may do it in a separate patch, it is not related to this one.

olavphibm commented on lib/config.py:
```diff
@@ -149,46 +49,61 @@ def __init__(self):
+            for attribute in config_metadata[section]:
+                argument = attribute.replace("_", "-")
+                argument_default = config_metadata[section][attribute]["default"]
+                argument_type = type(argument_default)
```
This will cause problems for arguments with `None` defaults.

alanoe replied:
No, it will not. Talked to Olav personally about this and he agreed

olavphibm commented on lib/config.py:
```diff
@@ -149,46 +49,61 @@ def __init__(self):
+                target_parser = self.parser
+            config_metadata[section].pop("help")
+            for attribute in config_metadata[section]:
+                argument = attribute.replace("_", "-")
```
Should we keep the hyphens instead of underscores in command line? This can be confusing when comparing with the config file.

alanoe replied:
Linux command line arguments usually use "-" as separator, but up to you guys.

olavphibm commented on lib/config.py:
```diff
@@ -149,46 +49,61 @@ def __init__(self):
+                    else:
+                        action= "store_false"
+                elif argument_type is list:
+                        keyword_args["nargs"] = "*"
```
Wrong indentation.

alanoe replied:
Done

olavphibm commented on lib/config.py:
```diff
@@ -149,46 +49,61 @@ def __init__(self):
+                keyword_args = {"help": argument_help}
+                if argument_type is bool:
+                    if bool(argument_default):
+                        argument = "--no-%s" % argument[2:]
```
This requires setting the `dest` option to `attribute`

alanoe replied:
Done

mopsfelder commented:
@alanoe As we discussed, we should avoid stacked pull requests as possible as we can.  At least, can you please state in the description which pull request this depends on?  For example:

```
Depends on #N
```

mopsfelder reviewed (changes requested):
Please update pull request description to properly track of dependencies.

alanoe commented:
MO> As we discussed, we should avoid stacked pull requests as possible as we can. At least, can you please state in the description which pull request this depends on?

This PR was created before we discussed that. Updating PR description.

Commit:
Remove comment about moving config.yaml to /etc

It was recently decided by the team that all build system files
should be inside the builds git repository. The motivation of
this is to make the build system more auto-contained and
be easily usable by several users in the same system.

Given this decision, it does not make sense to move the file to
/etc anymore.

olavphibm commented on lib/config.py:
```diff
@@ -149,46 +49,61 @@ def __init__(self):
+                        argument = "--no-%s" % argument[2:]
+                        action = "store_false"
+                    else:
+                        action= "store_false"
```
"store_true"

alanoe replied:
Done

alanoe commented:
LA> We need some documentation related to the config metadata file and the execution of sync_config_yaml.py.

Where this would go? CONTRIBUTING.md?


FR> My suggestion is to explicitly set these values in config_metadata now that this won't be edited by the user

This will make the config_metadata.yaml file unnecessarily *huge* ; I disagree, but whatever you guys prefer. However, this is not essential to this feature, so I'll put it in a separate PR.

alanoe commented:
start build

olavphibm commented on lib/config.py:
```diff
@@ -178,50 +60,81 @@ class ConfigParser(object):
+            target_parser = _subparsers[command]
+            for attribute in config_metadata[section]:
+                if attribute == "help":
+                    continue
```
You don't need this, you've already popped "help" above.

mopsfelder commented on lib/config.py:
```diff
@@ -294,14 +295,22 @@ def parse(self, command_line_args=None):
+                        config[node_name][key] = value
+                    elif type(config[node_name][key]) == dict:
+                        config[node_name][key] = json.loads(value)
                 command_line_args.pop(key)
```
How integer values, like `log_size`, would be handled?

What if we abstracted this assignment logic to avoid code duplication?  For example:

```python
def assign_config_value(name, value):
  value_type = type(name)
  if value_type == dict:
    name = json.loads(value)
  else:
    name = value
```

Then `assign_config_value` could be used:

```python
for key, value in command_line_args.items():
  if key in config[node_name]:
    if value is not None:
      assign_config_value(config[node_name][key], value)
    command_line_args.pop(key)
```

@olavphibm @fabianorosas What do you think?

olavphibm replied:
I agree in having the function, however it needs some modifications to actually modify the dict:
```python
def assign_config_value(config_dict, key, value):
  value_type = type(config_dict[key])
  if value_type == dict:
    config_dict[key] = json.loads(value)
  else:
    config_dict[key] = value
```

mopsfelder commented on lib/build_manager.py:
```diff
@@ -44,8 +45,11 @@ def _build_packages(self, distro, packages):
+                "%s-%s-%s.cfg" % (distro.name, distro.version, distro.architecture))
+            if not os.path.isfile(mock_config_file):
+                raise exception.BaseException("Mock config file not found at %s" % mock_config_file)
             package_builder = mockbuilder.Mock(mock_config_file)
```
This is pinning the configuration file path to a value user cannot change, making it difficult for user to build using different configuration files.

👎

mopsfelder commented on lib/iso_builder.py:
```diff
@@ -46,8 +46,12 @@ def _run_mock_command(self, cmd):
+            "build-iso-%s-%s-%s.cfg" % (distro.name, distro.version, distro.architecture))
+        if not os.path.isfile(mock_config_file):
+            raise exception.BaseException("Mock config file not found at %s" % mock_config_file)
+
```
Same here.

👎

olavph commented:
Superseded by https://github.com/open-power-host-os/builds/pull/251
